### PR TITLE
Add client credential auth method

### PR DESCRIPTION
**Pull request by dirkkul**
This adds a new oidc authentication method to the client and refactors the authentication handling

You can use this authentication mode with
```
client = weaviate.Client(
        url=URL, auth_client_secret=weaviate.AuthClientCredentials(client_secret=CLIENT_SECRET, scope=SCOPE),
    )
```

Tested with okta and azure

### CODE REVIEW

File: mock_tests/test_connection.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+import weaviate
+
+
+@pytest.mark.parametrize(
+    "header",
+    [
+        {},
+        {"Authorization": "Bearer test"},
+        {"Authorization": "Bearer test", "SomethingElse": "Value"},
```
Comment (StefanBogdan) [inline mock_tests/test_connection.py:15]: It should be case in-sensitive, so maybe you can add some tests to test for that too

File: weaviate/auth.py [hunk 2]
```diff
@@ -1,102 +1,38 @@
...
-        """
-        Get decoded credentials.
 
-        Returns
-        -------
-        dict
-            Decoded credentials.
-        """
+@dataclass
+class AuthBearerConfig(AuthCredentials):
```
Comment (StefanBogdan) [inline weaviate/auth.py:35]: This seems to be the same as `additional_headers`'s `Authorization` token. Do we want to have both of them?

**Comment by StefanBogdan** [comment]
Do we want to add same test for Google authentication?

**Comment by dirkkul** [comment]
> Do we want to add same test for Google authentication

Google sadly doesn't support client credentials. I haven't tested user/pw, but I'll make sure that the next method is supported by google

File: weaviate/auth.py (hunk 2 above)
Comment (dirkkul) [reply weaviate/auth.py:35]: yep, I thought it is more userfrendly to put it at the same place as the other authorization methods, but didn't want to break any existing setups. Do you have an opinion here? :)

File: mock_tests/test_connection.py (hunk 1 above)
Comment (dirkkul) [reply mock_tests/test_connection.py:15]: Hmm, this test only makes sure that what you set as additional headers shows up in the actual http-header. Which part should be case-insensitive?

Comment (StefanBogdan) [reply mock_tests/test_connection.py:15]: the `key` (this is true for all HTTP headers): https://github.com/semi-technologies/weaviate-python-client/blob/main/weaviate/connect/connection.py#L85-L86

File: weaviate/auth.py (hunk 2 above)
Comment (StefanBogdan) [reply weaviate/auth.py:35]: I think we should have both options for now.

Comment (dirkkul) [reply weaviate/auth.py:35]: oki, then let's leave it :)

File: mock_tests/test_connection.py (hunk 1 above)
Comment (dirkkul) [reply mock_tests/test_connection.py:15]: added an additional test with lowercase authentication

**Pull request merged**